play-scala-secure-session-example/app/services/session/SessionCache.scala

```diff
@@ -66,6 +66,15 @@ class SessionCache(
     case InternalSessionActorTerminated(key) =>
       behavior(children - key)
 
+    case _: InternalGetResponse =>
+      // upgrading to scala 2.13.4 required adding this case to prevent the compiler error due
+      // because the match was not exhaustive
+      // [error] ./play-scala-secure-session-example/app/services/session/SessionCache.scala:48:51: match may not be exhaustive.
+      // [error] It would fail on the following inputs: InternalGetResponse(GetDataDeleted(_, _), _), InternalGetResponse(GetFailure(_, _), _), InternalGetResponse(GetSuccess(_, _), _), InternalGetResponse(NotFound(_, _), _)
+      // [error]   ): Behavior[Command] = Behaviors.receiveMessage {
+      // [error]                                                   ^
+      Behaviors.same
+
```

Comment:
This is the most relevant change. I didn't spend time on reviewing all possible cases and just added a _catch-all_ new case to make the match exhaustive.

Reply:
@dwijnand I'd like your opinion on this...

Reply:
I am tempted to write the code so `InternalGetResponse` doesn't use an Akka Distributed Data trait on the first field (which causes this pattern matching explosion).

Reply:
The error is in how the `unapply`s are defined:

```scala
  object GetSuccess {
    def unapply[A <: ReplicatedData](rsp: GetSuccess[A]): Option[Key[A]] = Some(rsp.key)
  }
  object NotFound {
    def unapply[A <: ReplicatedData](rsp: NotFound[A]): Option[Key[A]] = Some(rsp.key)
  }
```

Those should return `Some` instead.  Doing that would get rid of the spurious `GetSuccess` and `NotFound` cases.  `GetFailure` and `GetDataDeleted` seems like they're legitimately not handled.  Go exhaustivity checker.

As scalac 2.13.4+ doesn't disable the exhaustivity checker with custom extractors it means you can write and use a locally-implemented workaround, something like:

```scala
  object FixedGetSuccess {
    def unapply[A <: ReplicatedData](rsp: GetSuccess[A]): Some[Key[A]] = Some(rsp.key)
  }
  object FixedNotFound {
    def unapply[A <: ReplicatedData](rsp: NotFound[A]): Some[Key[A]] = Some(rsp.key)
  }
```

Reply:
Thanks @dwijnand for the detailed explanation but I think I'm missing something. When I try to use the custom extractors (without the catch-all case) I still see too many cases reported as missing:

```scala
  object FixedGetSuccess {
    def unapply[A <: ReplicatedData](rsp: GetSuccess[A]): Some[Key[A]] = Some(rsp.key)
  }
  object FixedNotFound {
    def unapply[A <: ReplicatedData](rsp: NotFound[A]): Some[Key[A]] = Some(rsp.key)
  }

  def behavior(children: Map[String, ActorRef[RefreshSession.type]]): Behavior[Command] = Behaviors.receiveMessage {
    case PutInCache(key, value) =>
      replicator.askUpdate(Update(dataKey(key), emptyMap, WriteLocal, _)(_ :+ (key -> value)), nop)
      refreshSessionExpiration(key, children)

    case Evict(key) =>
      destroySessionExpiration(key)
      replicator.askUpdate(Update(dataKey(key), emptyMap, WriteLocal, _)(_.remove(uniqAddress, key)), nop)
      Behaviors.same

    case GetFromCache(key, replyTo) =>
      replicator.askGet(Get(dataKey(key), ReadLocal, _), InternalGetResponse(_, replyTo))
      Behaviors.same

    case InternalGetResponse(g @ FixedGetSuccess(mk @ LWWMapKey(key)), replyTo) =>
      replyTo ! Cached(key, g.get(mk).get(key))
      refreshSessionExpiration(key, children)

    case InternalGetResponse(FixedNotFound(LWWMapKey(key)), replyTo) =>
      replyTo ! Cached(key, None)
      Behaviors.same

    case InternalSessionActorTerminated(key) =>
      behavior(children - key)

//    case _: InternalGetResponse =>
//      Behaviors.same

    case _: InternalUpdateResponse[_] =>
      Behaviors.same
```

Produces:

```
[error] It would fail on the following inputs: 
InternalGetResponse(GetDataDeleted(_, _), _), 
InternalGetResponse(GetFailure(_, _), _), 
InternalGetResponse(GetSuccess(_, _), _), 
InternalGetResponse(NotFound(_, _), _)
[error]   def behavior(children: Map[String, ActorRef[RefreshSession.type]]): Behavior[Command] = Behaviors.receiveMessage {
[error]                                                                
```

I would expect the error to only list:

```
[error] It would fail on the following inputs: 
InternalGetResponse(GetDataDeleted(_, _), _), 
InternalGetResponse(GetFailure(_, _), _), 
[error]   def behavior(children: Map[String, ActorRef[RefreshSession.type]]): Behavior[Command] = Behaviors.receiveMessage {
[error]                                                                
```

Reply:
That's because LWWMapKey is only one of many ReplicatedData that GetSuccess and NotFound can contain.

Reply:
I just simplified back to only adding;

```
    case _: InternalGetResponse =>
      Behaviors.same
```

I don't think it's the goal of this sample to teach users how to pattern match in a DData structure. It'd be great but we should focus.